woocommerce/woocommerce-ios
# [REST API] Alamofire network retry and adapt network requests

Closes: #8482


## Description
As part of the REST API project we need the capability to retry a request in case of authentication failure. 

The flow for a REST API request is described below,
1. Send a request
2. If an error occurs due to authentication failure for a REST request. 
3. Regenerate a new application password.
4. Retry the REST API request again.

The request adapt and retry implementation is followed from https://github.com/wordpress-mobile/WordPressKit-iOS/blob/trunk/WordPressKit/Authenticator.swift

There are no changes to how a Jetpack or WPCOM request is handled. They will be authenticated using the new `RequestAuthenticator`. But there will be no retrying. 

### UI tests
As the network started validating the requests using Alamofire's `validate` method [like this](https://github.com/woocommerce/woocommerce-ios/pull/8483/files#diff-e5de4f95ce59918843a27539494c0fe4cb56643666b6ba692a8598eadbb1a35cR60), the UI tests started failing for mocks without `headers` section. 

The UI tests were failing due to incorrect `Content-Type` in the response.  The `headers` section takes care of adding the `Content-Type` to the mock response. 

To fix this, I added the headers part for all missing mock JSON files. 

## Testing instructions
- Login into the app using WPCOM credentials. (Both Email and Store Address flows)
- Do smoke testing to ensure that everything works as before.
- We don't have any REST API requests migrated. So there is no testing required. Coupons endpoints will be migrated in a future PR. 

## Screenshots
NA


---
